experimental/list_formatter/src/deduplicating_array.rs

```diff
@@ -0,0 +1,253 @@
+}
+
+pub fn deserialize<'de, D, T, const N: usize>(deserializer: D) -> Result<[T; N], D::Error>
+where
+    D: Deserializer<'de>,
+    T: Deserialize<'de> + Clone,
+    [HumanDe<T>; N]: Deserialize<'de>,
+    [MachineDe<T>; N]: Deserialize<'de>,
+{
+    use core::mem::MaybeUninit;
+    use serde::de::Error;
+
+    let mut array: [MaybeUninit<T>; N] = unsafe { MaybeUninit::uninit().assume_init() };
+
+    if deserializer.is_human_readable() {
+        for (i, r) in
+            IntoIterator::into_iter(<[HumanDe<T>; N]>::deserialize(deserializer)?).enumerate()
+        {
+            match r {
+                HumanDe::Value(v) => {
+                    array[i].write(v);
+                }
+                HumanDe::Fallback([j]) => unsafe {
+                    // Fallbacks should always be to a previous value,
+                    // which makes the assume_init_ref safe
+                    if j >= i {
+                        return Err(D::Error::custom(alloc::format!(
+                            "Illegal forward fallback {}->{}",
+                            i,
+                            j
+                        )));
+                    }
+                    array[i].write(array[j].assume_init_ref().clone());
+                },
+            }
+        }
+    } else {
+        for (i, r) in
+            IntoIterator::into_iter(<[MachineDe<T>; N]>::deserialize(deserializer)?).enumerate()
+        {
+            match r {
+                MachineDe::Value(v) => {
+                    array[i].write(v);
+                }
+                MachineDe::Fallback(j) => unsafe {
+                    // Fallbacks should always be to a previous value,
+                    // which makes the assume_init_ref safe
+                    if j >= i {
+                        return Err(D::Error::custom(alloc::format!(
+                            "Illegal forward fallback {}->{}",
+                            i,
+                            j
+                        )));
+                    }
+                    array[i].write(array[j].assume_init_ref().clone());
+                },
+            }
+        }
+    }
+
+    // https://github.com/rust-lang/rust/issues/61956
+    Ok(unsafe { core::ptr::read(array.as_ptr() as *const [T; N]) })
+}
+
+#[derive(serde::Serialize)]
+#[serde(untagged)]
+enum HumanSer<'a, T> {
+    Value(&'a T),
+    Fallback([usize; 1]),
+}
+
+enum MachineSer<'a, T> {
+    Value(&'a T),
+    Fallback(usize),
+}
+
+// These enums are public because they are exposed in the trait bounds of
+// dedupe_array_de. Serde only implements Deserialize<'de> on arrays up to
+// size 32, so [Dedupe<T>;N]: Deserialize<'de> cannot be inferred from
+// Dedupe<T>: Deserialize<'de> in the general case. See
+// https://github.com/serde-rs/serde/issues/1937.
+//
+// These are not considered part of the stable API.
+
+#[derive(serde::Deserialize)]
+#[serde(untagged)]
+#[doc(hidden)]
+pub enum HumanDe<T> {
+    Value(T),
+    Fallback([usize; 1]),
+}
```

Comment:
Nit/Suggestion: I think `HumanSer` / `HumanDe` are probably overkill, and you should use `MachineSer` / `MachineDe` for both cases.  The code would be more straightforward if you got rid of `HumanSer` / `HumanDe`, and it's not clear what you really gain from this separation.  If you let JSON do the externally tagged representation, as it should do with the machine versions, you would get something like

```json
[
  {
    "Value": {
      "default": {
        "string": ", ",
        "index_1": 2
      },
      "special_case": null
    }
  },
  {
    "Fallback": 0
  },
  {
    "Value": {
      "default": {
        "string": ", o ",
        "index_1": 4
      },
      "special_case": null
    }
  },
  {
    "Value": {
      "default": {
        "string": " o ",
        "index_1": 3
      },
      "special_case": null
    }
  },
  {
    "Fallback": 0
  }
]
```

I think this is not much worse than what you currently have:

```json
[
  {
    "default": {
      "string": ", ",
      "index_1": 2
    },
    "special_case": null
  },
  [
    0
  ],
  {
    "default": {
      "string": ", o ",
      "index_1": 4
    },
    "special_case": null
  },
  {
    "default": {
      "string": " o ",
      "index_1": 3
    },
    "special_case": null
  },
  [
    0
  ]
]
```

Reply:
The bincode encoding actually maps to this JSON
```json
[
  {
    "Value": {
      "default": {
        "string": ", ",
        "index_1": 2
      },
      "special_case": null
    }
  },
  {
    "Fallback: 0": {}
  },
  {
    "Value": {
      "default": {
        "string": ", o ",
        "index_1": 4
      },
      "special_case": null
    }
  },
  {
    "Value": {
      "default": {
        "string": " o ",
        "index_1": 3
      },
      "special_case": null
    }
  },
  {
    "Fallback: 0": {}
  }
]
```
i.e. the fallback is encoded in the variant (this lets us always save a byte, and support `usize` indices with varint encoding). I think this looks pretty bad, but in any case it's not technically possible, as the variant string has to be `&'static str`. And if we need a separate formatter for JSON anyway, we might as well make it look nice.

Reply:
Nit: I still prefer a tagged JSON representation.